Lib/test/test_urllib.py

```diff
@@ -1463,6 +1463,28 @@ def open_spam(self, url):
                 "spam://c:|windows%/:=&?~#+!$,;'@()*[]|/path/"),
                 "//c:|windows%/:=&?~#+!$,;'@()*[]|/path/")
 
+    def test_urlopener_retrieve_file(self):
+        with self.assertWarns(DeprecationWarning):
+            try:
+                fd, tmp_file = tempfile.mkstemp()
+                fileurl = 'file:' + urllib.request.pathname2url(tmp_file)
+                filename, _ = urllib.request.URLopener().retrieve(fileurl)
+                self.assertEqual(filename, tmp_file)
+            finally:
+                os.close(fd)
+                os.unlink(tmp_file)
+
+    def test_urlopener_retrieve_remote(self):
+        with self.assertWarns(DeprecationWarning):
+            try:
+                self.fakehttp(b"HTTP/1.1 200 OK\r\n\r\nHello!")
+                url = "http://www.python.org/file.txt"
+                filename, _ = urllib.request.URLopener().retrieve(url)
+                self.assertEqual(os.path.splitext(filename)[1], '.txt')
+            finally:
+                self.unfakehttp()
```

Comment:
Use of `self.addCleanup(self.unfakehttp)` should help to get rid of the whole try...finally block.